README.md
Cron-Sentry: error reporting to [Sentry](https://getsentry.com/) of commands run via cron
================================================

Cron-Sentry is a python command-line wrapper that reports errors to [Sentry](http://getsentry.com) (using [raven](https://github.com/getsentry/raven-python)) if the called script exits with a status other than zero.

Install
-------

`pip install cron-sentry`

Usage
-----

```
$ cron-sentry --help
usage: cron-sentry [-h] [--dsn SENTRY_DSN] [-M STRING_MAX_LENGTH] [--quiet] [--version] cmd [arg ...]

Wraps commands and reports those that fail to Sentry.

positional arguments:
  cmd                   The command to run

optional arguments:
  -h, --help            show this help message and exit
  --dsn SENTRY_DSN      Sentry server address
  -M STRING_MAX_LENGTH, --string-max-length STRING_MAX_LENGTH, --max-message-length STRING_MAX_LENGTH
                        The maximum characters of a string that should be sent
                        to Sentry (defaults to 4096)
  -q, --quiet           suppress all command output
  --version             show program's version number and exit

The Sentry server address can also be specified through the SENTRY_DSN
environment variable (and the --dsn option can be omitted).
```

It's possible to send extra information to Sentry via environment variables (prefix them with `CRON_SENTRY_EXTRA_`), such as:

```
$ export CRON_SENTRY_EXTRA_env=production
$ cron-sentry my-program
```

Example
-------

`crontab -e`
```
SENTRY_DSN=https://<your_key>:<your_secret>@app.getsentry.com/<your_project_id>
0 4 * * * cron-sentry my-process --arg arg2
```


License
-------

This project started life as [raven-cron](https://github.com/mediacore/raven-cron) by MediaCore Technologies.

Original copyright 2013 to MediaCore Technologies (MIT license).
Copyright 2015 to Yipit Inc. (MIT license).


CHANGELOG.md
current
=======

0.7.0
=====

* Send extra data to Sentry. Any `CRON_SENTRY_EXTRA_` prefixed env var is sent to Sentry (https://github.com/Yipit/cron-sentry/pull/18)


0.6.5
=====

* Send OSError message to Sentry (refs https://github.com/Yipit/cron-sentry/issues/3)


0.6.4
=====

* Send OSError exceptions to Sentry. Fix https://github.com/Yipit/cron-sentry/issues/3


0.6.3
=====

* Change HTTPTransport import to be compatible with old versions of raven-python


0.6.2
=====

* Use Raven's synchronous blocking transport (HTTPTransport)


0.6.1
=====

* Default `string_max_length` to 4096 and pass it to `raven.Client` (https://github.com/Yipit/cron-sentry/pull/14 and https://github.com/Yipit/cron-sentry/issues/10)


0.6.0
=====

* Require `argparse` in setup.py if running on Python 2.6 or older
* Add Python 3 compatibility


0.5.1
=====

* Add parameter --quiet for suppressing cmd output.


0.5.0
=====

* Replace `string-max-length` by `max-message-length`/`-M`. This is not compatible with version 0.4.4.


0.4.4
======

* Make the `cmd` parameter required (https://github.com/Yipit/cron-sentry/issues/7)
* Preserve exit status code from the specified command (https://github.com/Yipit/cron-sentry/issues/4)
* Add parameter `string-max-length` to command line (https://github.com/Yipit/cron-sentry/issues/9)


0.4.3
=====

* Fix PyPI package to include README.md (https://github.com/Yipit/cron-sentry/pull/8)


0.4.2
=====

* Change argparse to use REMAINDER instead of '+' for `cmd`. More details at https://github.com/Yipit/cron-sentry/pull/6
* Make cron-sentry compatible with old and new way of specifying arguments in the command line
* Change usage message to have `cmd [arg ...]` rather than only `...`
* Update setup.py to use README.md as `long_description`


0.4.1
=====

* Upload right files to PyPI


0.4.0
=====

* Renamed fork from raven-cron to cron-sentry


0.3.x
=====
* Yipit forks the project and keep it under the same name
* Add support to command options (commit 75ade6d920cbb9f0e84575ca3ed8b568f945727d)
* Preserve stdout and stderr (commit a604f3e0a104ab018fd1c9b35fe989fdd0834a4b)
* Python 3 support, added by @Ian-Foote (commit a0b74c504e748a821fd07236204c52439cc1b1e9)


0.2.0 / 2013-12-18 
==================

 * Update RELEASE.md
 * How to make a release


0.1.0 / 2013-10-22 
==================

 * Fixes .gitignore
 * Don't load the raven dependency in `__init__.py`. Fixes #1
 * Project moved
 * Add LICENSE file



tests/test_command_reporter.py
import errno
import os

import mock
import sys

from cron_sentry.runner import CommandReporter, DEFAULT_STRING_MAX_LENGTH, run, parser


@mock.patch('cron_sentry.runner.Client')
def test_command_reporter_accepts_parameters(ClientMock):
    reporter = CommandReporter(['date', '--invalid-option'], 'http://testdsn', DEFAULT_STRING_MAX_LENGTH)

    reporter.run()

    client = ClientMock()
    assert client.captureMessage.called


@mock.patch('cron_sentry.runner.sys')
@mock.patch('cron_sentry.runner.Client')
def test_command_reporter_catches_invalid_commands(ClientMock, sys_mock):
    command = ['command-does-not-exists']
    reporter = CommandReporter(command, 'http://testdsn', DEFAULT_STRING_MAX_LENGTH)

    exit_code = reporter.run()

    expected_exit_code = 127
    expected_stdout = ''
    expected_stderr = '[Errno {errno}] {msg}'.format(errno=errno.ENOENT, msg=os.strerror(errno.ENOENT))

    client = ClientMock()
    client.captureMessage.assert_called_with(
        mock.ANY,
        time_spent=mock.ANY,
        data=mock.ANY,
        extra={
            'command': command,
            'exit_status': expected_exit_code,
            "last_lines_stdout": expected_stdout,
            "last_lines_stderr": mock.ANY,
    })
    assert exit_code == expected_exit_code
    sys_mock.stdout.write.assert_called_with(expected_stdout)
    # python 3 exception `FileNotFoundError` has additional content compared to python 2's `OSError`
    assert client.captureMessage.call_args[1]['extra']['last_lines_stderr'].startswith(expected_stderr)
    assert sys_mock.stderr.write.call_args[0][0].startswith(expected_stderr)


@mock.patch('cron_sentry.runner.Client')
def test_command_reporter_works_with_no_params_commands(ClientMock):
    reporter = CommandReporter(['date'], 'http://testdsn', DEFAULT_STRING_MAX_LENGTH)

    reporter.run()

    client = ClientMock()
    assert not client.captureMessage.called


@mock.patch('cron_sentry.runner.sys')
@mock.patch('cron_sentry.runner.Client')
def test_command_reporter_keeps_stdout_and_stderr(ClientMock, sys_mock):
    command = [sys.executable, '-c', """
import sys
sys.stdout.write("test-out")
sys.stderr.write("test-err")
sys.exit(2)
"""]
    reporter = CommandReporter(command, 'http://testdsn', DEFAULT_STRING_MAX_LENGTH)
    client = ClientMock()

    reporter.run()

    sys_mock.stdout.write.assert_called_with('test-out')
    sys_mock.stderr.write.assert_called_with('test-err')
    client.captureMessage.assert_called_with(
        mock.ANY,
        time_spent=mock.ANY,
        data=mock.ANY,
        extra={
            'command': command,
            'exit_status': 2,
            "last_lines_stdout": "test-out",
            "last_lines_stderr": "test-err",
    })


@mock.patch('cron_sentry.runner.sys')
@mock.patch('cron_sentry.runner.Client')
def test_reports_correctly_to_with_long_messages_but_trims_stdout_and_stderr(ClientMock, sys_mock):
    command = [sys.executable, '-c', """
import sys
sys.stdout.write("a" * 20000)
sys.stderr.write("b" * 20000)
sys.exit(2)
"""]
    reporter = CommandReporter(command, 'http://testdsn', DEFAULT_STRING_MAX_LENGTH)
    client = ClientMock()

    reporter.run()
    expected_stdout = '...{0}'.format('a' * (DEFAULT_STRING_MAX_LENGTH - 3))
    expected_stderr = '...{0}'.format('b' * (DEFAULT_STRING_MAX_LENGTH - 3))

    sys_mock.stdout.write.assert_called_with(expected_stdout)
    sys_mock.stderr.write.assert_called_with(expected_stderr)
    client.captureMessage.assert_called_with(
        mock.ANY,
        time_spent=mock.ANY,
        data=mock.ANY,
        extra={
            'command': command,
            'exit_status': 2,
            "last_lines_stdout": expected_stdout,
            "last_lines_stderr": expected_stderr,
    })


@mock.patch('cron_sentry.runner.sys')
@mock.patch('cron_sentry.runner.CommandReporter')
def test_command_line_should_support_command_args_without_double_dashes(CommandReporterMock, sys_mock):
    command = ['--dsn', 'http://testdsn', 'command', '--arg1', 'value1', '--arg2', 'value2']

    run(command)

    CommandReporterMock.assert_called_with(
        cmd=command[2:],
        dsn='http://testdsn',
        string_max_length=DEFAULT_STRING_MAX_LENGTH,
        quiet=False,
        extra={},
    )


@mock.patch('cron_sentry.runner.sys')
@mock.patch('cron_sentry.runner.CommandReporter')
def test_command_line_should_support_command_with_double_dashes(CommandReporterMock, sys_mock):
    command = ['--dsn', 'http://testdsn', '--', 'command', '--arg1', 'value1', '--arg2', 'value2']

    run(command)

    CommandReporterMock.assert_called_with(
        cmd=command[3:],
        dsn='http://testdsn',
        string_max_length=DEFAULT_STRING_MAX_LENGTH,
        quiet=False,
        extra={},
    )


@mock.patch('cron_sentry.runner.sys')
@mock.patch('argparse._sys')
@mock.patch('cron_sentry.runner.CommandReporter')
def test_should_display_help_text_and_exit_with_1_if_no_command_is_specified(CommandReporterMock, argparse_sys, cron_sentry_sys):
    command = []
    run(command)

    cron_sentry_sys.stderr.write.assert_called_with("ERROR: Missing command parameter!\n")
    argparse_sys.stdout.write.assert_called_with(parser.format_usage())
    cron_sentry_sys.exit.assert_called_with(1)
    assert not CommandReporterMock.called


@mock.patch('cron_sentry.runner.sys')
@mock.patch('cron_sentry.runner.Client')
def test_exit_status_code_should_be_preserved(ClientMock, sys_mock):
    command = [sys.executable, '-c', 'import sys; sys.exit(123)']

    run(command)

    sys_mock.exit.assert_called_with(123)



@mock.patch('cron_sentry.runner.sys')
@mock.patch('cron_sentry.runner.Client')
def test_should_trim_stdout_and_stderr_based_on_command_line(ClientMock, sys_mock):
    command = [
    '--dsn', 'http://testdsn',
    '--max-message-length', '100',
    sys.executable, '-c', """
import sys
sys.stdout.write("a" * 20000 + "end")
sys.stderr.write("b" * 20000 + "end")
sys.exit(2)
"""]

    run(command)

    # -3 refers to "..." and "end"
    expected_stdout = '...{0}end'.format('a' * (100 - 3 - 3))
    expected_stderr = '...{0}end'.format('b' * (100 - 3 - 3))

    sys_mock.stdout.write.assert_called_with(expected_stdout)
    sys_mock.stderr.write.assert_called_with(expected_stderr)
    client = ClientMock()
    client.captureMessage.assert_called_with(
        mock.ANY,
        time_spent=mock.ANY,
        data=mock.ANY,
        extra={
            'command': mock.ANY,
            'exit_status': mock.ANY,
            "last_lines_stdout": expected_stdout,
            "last_lines_stderr": expected_stderr,
    })

@mock.patch('cron_sentry.runner.sys')
@mock.patch('cron_sentry.runner.Client')
def test_should_suppress_stdout_and_stderr_based_on_command_line(ClientMock, sys_mock):
    command = [
    '--dsn', 'http://testdsn',
    '--quiet',
    sys.executable, '-c', """
import sys
sys.stdout.write("a" * 100 + "end")
sys.stderr.write("b" * 100 + "end")
sys.exit(2)
"""]

    run(command)

    expected_stdout = "a" * 100 + "end"
    expected_stderr = "b" * 100 + "end"

    assert not sys_mock.stdout.write.called
    assert not sys_mock.stderr.write.called

    client = ClientMock()
    client.captureMessage.assert_called_with(
        mock.ANY,
        time_spent=mock.ANY,
        data=mock.ANY,
        extra={
            'command': mock.ANY,
            'exit_status': mock.ANY,
            "last_lines_stdout": expected_stdout,
            "last_lines_stderr": expected_stderr,
    })


@mock.patch('cron_sentry.runner.sys')
@mock.patch('cron_sentry.runner.Client')
def test_extra_data_via_env_vars_should_go_to_sentry(ClientMock, sys_mock):
    command = ['--dsn', 'http://testdsn', 'false']

    os.environ['CRON_SENTRY_EXTRA_secret1'] = 'hello'
    os.environ['CRON_SENTRY_EXTRA_secret2'] = 'world'
    try:
        run(command)
    finally:
        del os.environ['CRON_SENTRY_EXTRA_secret1']
        del os.environ['CRON_SENTRY_EXTRA_secret2']

    expected_stdout = "a" * 100 + "end"
    expected_stderr = "b" * 100 + "end"

    client = ClientMock()
    client.captureMessage.assert_called_with(
        mock.ANY,
        time_spent=mock.ANY,
        data=mock.ANY,
        extra={
            'command': mock.ANY,
            'exit_status': mock.ANY,
            'last_lines_stdout': mock.ANY,
            'last_lines_stderr': mock.ANY,
            'secret1': 'hello',
            'secret2': 'world',
    })


cron_sentry/runner.py
import sys
from getpass import getuser
from os import getenv, path, SEEK_END
from raven import Client
from raven.transport import HTTPTransport
from subprocess import call
from tempfile import TemporaryFile
from argparse import ArgumentParser, REMAINDER
from sys import argv
from time import time
from .version import VERSION


# 4096 is more than Sentry will accept by default. SENTRY_MAX_EXTRA_VARIABLE_SIZE in the Sentry configuration
# also needs to be increased to allow longer strings.
DEFAULT_STRING_MAX_LENGTH = 4096


parser = ArgumentParser(
    description='Wraps commands and reports those that fail to Sentry.',
    epilog=('The Sentry server address can also be specified through ' +
            'the SENTRY_DSN environment variable ' +
            '(and the --dsn option can be omitted).'),
    usage='cron-sentry [-h] [--dsn SENTRY_DSN] [-M STRING_MAX_LENGTH] [--quiet] [--version] cmd [arg ...]',
)
parser.add_argument(
    '--dsn',
    metavar='SENTRY_DSN',
    default=getenv('SENTRY_DSN'),
    help='Sentry server address',
)

parser.add_argument(
    '-M', '--string-max-length', '--max-message-length',
    type=int,
    default=DEFAULT_STRING_MAX_LENGTH,
    help='The maximum characters of a string that should be sent to Sentry (defaults to {0})'.format(DEFAULT_STRING_MAX_LENGTH),
)
parser.add_argument(
    '-q', '--quiet',
    action='store_true',
    default=False,
    help='suppress all command output'
)
parser.add_argument(
    '--version',
    action='version',
    version=VERSION,
)
parser.add_argument(
    'cmd',
    nargs=REMAINDER,
    help='The command to run',
)

# subprocess.call() raises `OSError` in Python 2 but `FileNotFoundError` in Python 3.
# The FileNotFoundError exception is defined only in Python 3 and the following is a shim.
try:
    CommandNotFoundError = FileNotFoundError
except NameError:
    CommandNotFoundError = OSError


def update_dsn(opts):
    """Update the Sentry DSN stored in local configs

    It's assumed that the file contains a DSN endpoint like this:
    https://public_key:secret_key@app.getsentry.com/project_id

    It could easily be extended to override all settings if there
    were more use cases.
    """

    homedir = path.expanduser('~%s' % getuser())
    home_conf_file = path.join(homedir, '.cron-sentry')
    system_conf_file = '/etc/cron-sentry.conf'

    conf_precedence = [home_conf_file, system_conf_file]
    for conf_file in conf_precedence:
        if path.exists(conf_file):
            with open(conf_file, "r") as conf:
                opts.dsn = conf.read().rstrip()
            return


def run(args=argv[1:]):
    opts = parser.parse_args(args)

    # Command line takes precendence, otherwise check for local configs
    if not opts.dsn:
        update_dsn(opts)

    if opts.cmd:
        # make cron-sentry work with both approaches:
        #
        #     cron-sentry --dsn http://dsn -- command --arg1 value1
        #     cron-sentry --dsn http://dsn command --arg1 value1
        #
        # see more details at https://github.com/Yipit/cron-sentry/pull/6
        if opts.cmd[0] == '--':
            cmd = opts.cmd[1:]
        else:
            cmd = opts.cmd
        runner = CommandReporter(
            cmd=cmd,
            dsn=opts.dsn,
            string_max_length=opts.string_max_length,
            quiet=opts.quiet
        )
        sys.exit(runner.run())
    else:
        sys.stderr.write("ERROR: Missing command parameter!\n")
        parser.print_usage()
        sys.exit(1)


class CommandReporter(object):
    def __init__(self, cmd, dsn, string_max_length, quiet=False):
        self.dsn = dsn
        self.command = cmd
        self.string_max_length = string_max_length
        self.quiet = quiet

    def run(self):
        start = time()

        with TemporaryFile() as stdout:
            with TemporaryFile() as stderr:
                try:
                    exit_status = call(self.command, stdout=stdout, stderr=stderr)
                except CommandNotFoundError as exc:
                    last_lines_stdout = ''
                    last_lines_stderr = str(exc)
                    exit_status = 127  # http://www.tldp.org/LDP/abs/html/exitcodes.html
                else:
                    last_lines_stdout = self._get_last_lines(stdout)
                    last_lines_stderr = self._get_last_lines(stderr)

                if exit_status != 0:
                    elapsed = int((time() - start) * 1000)
                    self.report_fail(exit_status, last_lines_stdout, last_lines_stderr, elapsed)

                if not self.quiet:
                    sys.stdout.write(last_lines_stdout)
                    sys.stderr.write(last_lines_stderr)

                return exit_status

    def report_fail(self, exit_status, last_lines_stdout, last_lines_stderr, elapsed):
        if self.dsn is None:
            return

        message = "Command \"%s\" failed" % (self.command,)

        client = Client(transport=HTTPTransport, dsn=self.dsn, string_max_length=self.string_max_length)

        client.captureMessage(
            message,
            data={
                'logger': 'cron',
            },
            extra={
                'command': self.command,
                'exit_status': exit_status,
                'last_lines_stdout': last_lines_stdout,
                'last_lines_stderr': last_lines_stderr,
            },
            time_spent=elapsed
        )

    def _get_last_lines(self, buf):
        buf.seek(0, SEEK_END)
        file_size = buf.tell()
        if file_size < self.string_max_length:
            buf.seek(0)
            last_lines = buf.read().decode('utf-8')
        else:
            buf.seek(-(self.string_max_length - 3), SEEK_END)
            last_lines = '...' + buf.read().decode('utf-8')
        return last_lines
